### ISSUE TITLE
Return >0 if the autosubmit command fails - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1878]
_Merges return-non-zero -> master_

Closes #1038 

Pending:

- [x] docs
- [ ] tests
- [x] manual tests using several AS commands as this merge request applies to multiple ones, not just `run`

**Comment by kinow** [comment]
added 1 commit

<ul><li>89b9b859 - Return &gt;0 if the autosubmit command fails</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/390/diffs?diff_id=18353&start_sha=5837b9686ea524f25a7457ce56748e3ef40689a3)

**Comment by kinow** [comment]
These commands all worked fine and resulted in exit codes `0` — we don't care about it returning errors in case of failures for now, the important command to do so for now is just `autosubmit run`.

- [x] autosubmit delete
- [x] autosubmit expid
- [x] autosubmit create
- [x] autosubmit refresh
- [x] autosubmit run (returns `>0` if failed jobs)
- [x] autosubmit report
- [x] autosubmit monitor
- [x] autosubmit stats
- [x] autosubmit archive
- [x] autosubmit unarchive
- [x] autosubmit updatedescrip

**Comment by kinow** [comment]
added 1 commit

<ul><li>1ca8810a - Add documentation about exit status of autosubmit run</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/390/diffs?diff_id=18447&start_sha=a67d86ba7fcdf0363e87ac9f6eddf60baabda9ff)

**Comment by kinow** [comment]
marked the checklist item **docs** as completed

**Comment by kinow** [comment]
Some quick docs written. Here's what they look like rendered locally:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/e5d93ce80d30390814e0e7197e1e675c/image.png)

**Comment by kinow** [comment]
Leaving the unit tests for tomorrow, then it should be ready for review :thumbsup:

**Comment by kinow** [comment]
marked this merge request as **ready**

**Comment by kinow** [comment]
requested review from @dbeltrankyl

**Comment by kinow** [comment]
Hi @dbeltrankyl , I was looking at the code today, planning a unit test, but I realized I would need to mock a lot of call to cover success and failure scenarios. I will create a separate issue for a list of functional tests to be written, including one for this issue :thumbsup: So this one should be ready to review. Thanks!

**Comment by kinow** [comment]
And set the milestone to 4.0, but feel free to change it to 4.1 if that's the next release (I think 4.1 is waiting on your merge request with performance improvements to be reviewed/merged).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 15, 2023, 10:48

Thanks Bruno, the code is okay.

One question, did you try to control + c ? It should also return 1, no?

If it is right we can close this as you mention that the test will be in a different merge

**Comment by kinow** [comment]
Huh, I didn't test that! :astonished: 

>It should also return 1, no?

~Yup~! Let me check if it does that.

Actually not sure now.

**Comment by kinow** [comment]
Alright, we confirmed (Slack) that this behavior should be fine.

Whenever the user hits CTRL C, Autosubmit will stop in the next iteration of the loop, and if there are no failed jobs it will stop cleanly and return `0`. Otherwise, if there are failures, it will return `1`.

i.e. an interrupted workflow execution returns `0` unless there are failed tasks, which sounds fair. And for An-Chi and Oriol that should not matter, as I believe they are calling it in non-interactive mode anyway when they want the exit status.

**Comment by kinow** [comment]
@dbeltrankyl created an issue to gather all the cases that need integration or functional tests, so that we have a list to work on once we have time for that https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1153

With that I think this MR may be approved?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 15, 2023, 11:24

approved this merge request

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 15, 2023, 11:24

Yes, thanks Bruno

**Comment by kinow** [comment]
Thank **you** Dani! Merging now...

**Comment by kinow** [comment]
mentioned in commit 9b98f1687924e53a9975007142bec57c498fa8a7
